payu/fsops.py

```diff
@@ -229,3 +234,89 @@ def list_archive_dirs(archive_path: Union[Path, str],
 
     dirs.sort(key=lambda d: int(d.lstrip(dir_type)))
     return dirs
+
+
+def run_script_command(script_cmd: str, control_path: Path) -> None:
+    """Run a user defined script or command.
+
+    Parameters
+    ----------
+    script_cmd : string
+        String of user-script command defined in configuration file
+    control_path : Path
+        The control directory of the experiment
+
+    Raises
+    ------
+    RuntimeError
+        If there's was an error running the user-script
+    """
+    try:
+        _run_script(script_cmd, control_path)
+    except Exception as e:
+        error_msg = f"User defined script/command failed to run: {script_cmd}"
+        raise RuntimeError(error_msg) from e
+
+
+def needs_subprocess_shell(command: str) -> bool:
+    """Check if command contains shell specific values. For example, file
+    redirections, pipes or logical operators.
+
+    Parameters
+    ----------
+    command: string
+        String of command to run in subprocess call
+
+    Returns
+    -------
+    bool
+        Returns True if command requires a subprocess shell, False otherwise
+    """
+    shell_values = ['>', '<', '|', '&&', '$', '`']
+    for value in shell_values:
+        if value in command:
+            return True
+    return False
+
+
+def _run_script(script_cmd: str, control_path: Path) -> None:
+    """Helper recursive function to attempt running a script command.
+
+    Parameters
+    ----------
+    script_cmd : string
+        The script command to attempt to run in subprocess call
+    control_path: Path
+        The control directory to use for resolving relative filepaths, if file
+        is not found
+    """
+    # First try to interpret the argument as a full command
+    try:
+        if needs_subprocess_shell(script_cmd):
+            subprocess.check_call(script_cmd, shell=True)
+        else:
+            subprocess.check_call(shlex.split(script_cmd))
+            print(script_cmd)
+
+    except FileNotFoundError:
+        # Check if script is a file in the control directory
+        cmd = control_path / script_cmd
+        if cmd.is_file():
+            _run_script(str(cmd), control_path)
+        else:
+            raise
+
+    except PermissionError:
+        # Guess the type of interpreter using the file extension
+        _, file_ext = os.path.splitext(script_cmd)
+        shell_name = EXTENSION_TO_INTERPRETER.get(file_ext, None)
+        if shell_name:
+            print('payu: warning: Assuming that {0} is a {1} '
+                  'script based on the filename extension.'
+                  ''.format(os.path.basename(script_cmd),
+                            os.path.basename(shell_name)))
```

Comment:
For minor styling, is there a preference for _f-strings_ in the project?

Also, is this OK as a `print` or would it be better in `warnings.warn()` call?

Reply:
I might keep as print for now and revisit it as part of this issue which aims to fix some of the warning formatting in the project  https://github.com/payu-org/payu/issues/451